### PR TITLE
Allow 24 nominations per validator and publish count

### PR DESCRIPTION
**Pull request by gavofyork**
polkadot companion: https://github.com/paritytech/polkadot/pull/2417

### CODE REVIEW

**Review by kianenigma — COMMENTED** [review]
I would take such direction to prevent all runtimes being bound to this configuration: 

Staking would accept `type CompactSolution: sp_npos_elections::CompactSolution<T::AccountId>` as input in which the max-naminations is implicitly defined. 

Later on, once #7909 is merged this configuration will be moved to the new election pallet and staking would only need a `type MaxNominations: Get<_>`.

**Comment by kianenigma** [comment]
Actually, doing the above requires a bit of making things generic in staking, which is pretty much in vain if we also merge #7909. I'd say lets wait until we are almost at the next release: 

if 7909 is merged, there's no need for this.
else, we merge this one. 

I'll get it in good shape anyhow though.

**Comment by kianenigma** [comment]
Should build well now and the 24 is configurable from the runtime. Lots of boilerplate. as you see, so I'd consider not doing it this way as this will all be removed by the linked PR anyhow.

File: frame/staking/src/lib.rs
Comment (kianenigma) [inline frame/staking/src/lib.rs:337]: Honestly I put this mostly here so that it can be blindly used in tests/benchmarks etc. Polkadot or any other chain should probably generate their own version.

File: frame/staking/src/lib.rs
Comment (kianenigma) [inline frame/staking/src/lib.rs:874]: TLDR ^^^^^^^^^^^^^^^^^

File: frame/offences/benchmarking/src/lib.rs
Comment (kianenigma) [inline frame/offences/benchmarking/src/lib.rs:212]: I kinda blindly changed these and didn't check the full scope, they were 16 before, and now they are 24, just different syntax.

**Comment by kianenigma** [comment]
Burnin looks well, will leave it running for another day with full logs to check everything, else we're good to go.

**Comment by gavofyork** [comment]
image

companion failing @kianenigma

File: frame/offences/benchmarking/src/lib.rs
Comment (thiolliere) [reply frame/offences/benchmarking/src/lib.rs:212]: we can reexecute benchmarks to see

File: frame/staking/src/lib.rs
Comment (thiolliere) [inline frame/staking/src/lib.rs:336]:
```suggestion
/// The default solution type used in substrate-node with 24 maximum votes per voter. A runtime can customize this to their needs.
```

**Review by thiolliere — COMMENTED** [review]
the title of the PR should be "Allow 24 nominations per nominator in election solution and publish count", no ?

File: frame/staking/src/lib.rs
Comment (thiolliere) [inline frame/staking/src/lib.rs:1406]: sidenote:

```suggestion
			// Ensure the size of both ValidatorIndex and NominatorIndex. They both need to be less or
			// equal to usize.
```
Also it would make more sense to compare their max_value. But I'm fine here

File: frame/staking/src/lib.rs
Comment (thiolliere) [inline frame/staking/src/lib.rs:2279]:
```suggestion
impl<T: Config> Module<T> {
```

**Review by thiolliere — APPROVED** [review]
staking weight should be updated also I think,

otherwise looks good to me

**Comment by kianenigma** [comment]
Has a companion now, locally my build was borked but had something to do with tracing in network crate, so def. not my issue. Will wait to see the result here. Either way, it is ready for audit.

**Comment by gnunicorn** [comment]
anyone working on this?

**Comment by kianenigma** [comment]
awaiting audit as it is no longer trivial.

**Comment by kianenigma** [comment]
re-intorduced in https://github.com/paritytech/substrate/pull/8113

**Pull request closed without merging**
